ffmpeg/ffmpeg.go

```diff
@@ -486,6 +488,253 @@ func Transcode3(input *TranscodeOptionsIn, ps []TranscodeOptions) (*TranscodeRes
 	return t.Transcode(input, ps)
 }
 
+// IMPORTANT: this function should be _always_ paired with corresponding
+// destroyCOutputParams call, even if error is returned. This is because we want
+// to make sure that we don't leak C memory. Old code used defer statement to
+// pair every C allocation with corresponding release, but then the results
+// lifetime was obviously limited to the function scope - not something desired
+// here. Alternative solution would be to create configuration in two passes,
+// one just for checking, another one for actual C memory allocation - in effect
+// making code more complex. Hence "always destroy" policy, and destroy function
+// is removing only these fields that actually _were_ initialized
+func createCOutputParams(input *TranscodeOptionsIn, ps []TranscodeOptions) ([]C.output_params, error) {
+    params := make([]C.output_params, len(ps))
+    for i, p := range ps {
```

Comment:
The alternative to `defer` is [SetFinalizer](https://pkg.go.dev/runtime#SetFinalizer) which is bad choice here.

What can be improvement is to return "finalizer" to the caller instead of relying on comments to remind us in the future. Go unused variable error will remind us to defer call to returned finalizer.

```suggestion
func createCOutputParams(input *TranscodeOptionsIn, ps []TranscodeOptions) ([]C.output_params, func(), error) {
    params := make([]C.output_params, len(ps))
    finalizer := func () { destroyCOutputParams(params) } // always return as second return value
    for i, p := range ps {
```

Reply:
Good idea @AlexKordic. Another option, and, possibly, more Golang way to implement this is with Context. As this logic ultimately runs in response to user request, there should be a Context associated with it, which could be passed to any routine willing to perform cleanup action:
```
func createCOutputParams(ctx Context, input *TranscodeOptionsIn, ps []TranscodeOptions) {
	defer func() {
		<-ctx.Done()
		destroyCOutputParams(params)
	}()
}
// somewhere up the call stack
ctx, cancel := context.WithCancel()
defer cancel();

// usage
createCOutputParams(ctx, ...)
```
We may not have uninterrupted Context 'flow' everywhere currently, so it could be created right in `configEncoder()` I think.

Reply:
One question about go nature: what is difference between `defer func() {<-ctx.Done()}()` and `go func() {<-ctx.Done()}()` ? My feeling is `defer` would block before `createCOutputParams` returns and i would expect `go` to not affect return, @cyberj0g is this correct?

Reply:
Thanks for your suggestions, guys. I read a bit about the Context approach and I really like it. However, I feel this code is going to change a lot in the future (for example I feel that this Nvidia specific filters stuff doesn't belong in high-level configuration code, but rather somewhere deep in pipeline implementation specialised for Nvidia HW). And thus I'd rather go with something simple. So I changed to what Alex suggested - I like that the compiler will remind the user should finaliser get forgotten.

Since there were no complaints I assume that I can turn that into ordinary PR.